core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocateUnassignedDecision.java

```diff
@@ -109,13 +111,7 @@ public AllocateUnassignedDecision(StreamInput in) throws IOException {
         allocationStatus = in.readOptionalWriteable(AllocationStatus::readFrom);
         assignedNode = in.readOptionalWriteable(DiscoveryNode::new);
         allocationId = in.readOptionalString();
-
-        Map<String, NodeAllocationResult> nodeDecisions = null;
-        if (in.readBoolean()) {
-            nodeDecisions = in.readMap(StreamInput::readString, NodeAllocationResult::new);
-        }
-        this.nodeDecisions = (nodeDecisions != null) ? Collections.unmodifiableMap(nodeDecisions) : null;
-
+        nodeDecisions = in.readBoolean() ? sortNodeDecisions(in.readMap(StreamInput::readString, NodeAllocationResult::new)) : null;
```

Comment:
I pushed 42aad8bf40c41cef68a689907bc5a6a07033fe7a which adds a `readMap` method to `StreamInput` that takes a map supplier.  I used the same code instead of having the older `readMap` method call into `readMap` with a hash map supplier because of the ability of the old `readMap` method to use the size param to properly size the map from the start.  In practice, this is likely not going to cost us much, so I'm also fine with having just one method that takes a map supplier and is not able to size the map from the beginning.   WDYT?